keps/sig-auth/3221-multiple-authorization-webhooks/README.md

```diff
@@ -0,0 +1,556 @@
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Today, `kube-apiserver` allows one to configure the authorization chain using a set of command line flags of the format `--authorization-*`.
+And cluster administrators can only enable one webhook as a part of the authorization chain using the `--authorization-modes` flag.
+This limits admins from creating authorization chains with multiple webhooks that validate requests in a certain order.
+This proposal makes the case for a more structured config to define the authorization chain for `kube-apiserver` and allowing multiple webhooks with well-defined parameters and enabling fine grained control like explicit Deny authorizer and more definitive `SubjectAccessReviews`.
+
+## Motivation
+
+Cluster administrators should be able to specify more than one authorization webhook in the API Server handler chain.
+They also need to be able to declaratively configure the authorizer chain using a configuration file.
+
+It should also be easy to say when to Deny requests, for example, when a webhook is unreachable.
+
+### Goals
+
+- Define a configuration file format for configuring Kubernetes API Server Authorization chain.
+- Allow ordered definition of authorization modes.
+- Allow definition of multiple webhooks in the authorization chain.
+- Enable user to define the policy when a webhook can't be reached due to network errors etc.
+- Enable more defnitive results for a `SubjectAccessReview` by allowing config to distinguish webhooks which might deny access.
+
+### Non-Goals
+
+- Allow resource/user based pre-filtering of webhooks to prevent unnecessary invocations.
+- Hot reload of the configuration.
+
+## Proposal
+
+Add a configuration format having specific precedence order and defined failure modes for configuring authorizer chain.
+
+```yaml
+apiVersion: apiserver.config.k8s.io/v1
+kind: AuthorizationConfiguration
+authorizers:
+- type: Webhook
+  webhook:
+    server: https://<path-to-webserver>
```

Comment:
Is this actually a URL?  Can it be like "https://foo.com/path/to/my/endpoint"?

Reply:
Yup. It can any path to a webserver accepting requests of the form that current Authz webhooks have.

The field name can surely be refined to something like `serverURL` or just `url`.